File: apps/android/app/src/main/java/com/voiid/app/store/UserDirectory.kt
package com.voiid.app.store

import android.content.Context
import androidx.compose.runtime.snapshots.SnapshotStateMap
import androidx.compose.runtime.mutableStateMapOf
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * The ONE place a human-readable name for a user id is produced — port of iOS
 * `UserDirectory.swift`.
 *
 * WHY: names were resolved ad hoc at every call site, each with its own fallback, and the
 * CALL paths had no fallback at all — so an incoming call showed a raw user UUID, and the
 * chat list showed whatever the peer typed at signup even when you had them saved as
 * something else. The address book WAS read correctly ([com.voiid.app.net.ContactsService]),
 * but only into a 120s in-memory cache that died on relaunch and was never consulted by the
 * call or chat UI.
 *
 * PRECEDENCE (applies everywhere — calls, chat list, chat header, profile):
 *   1. saved_name  — this device's address book. If you saved them as "Mum", every screen
 *                    says "Mum", whatever they call themselves.
 *   2. full_name   — the peer's own profile name.
 *   3. phone_e164  — a number is still useful.
 *   4. username    — Clips handle, last resort.
 *   5. "Unknown"   — NEVER a raw user id. A UUID on screen is always a bug.
 *
 * PRIVACY: this is also why the ring push carries no caller name. Sending one would tell
 * Google who is calling whom; instead the callee resolves the name locally from `caller_id`
 * through this directory, which needs no network and works when the app was just launched
 * cold by that push.
 */
object UserDirectory {

    /** Own IO scope: writes are fire-and-forget from UI/signaling threads. */
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    @Volatile private var dao: UserDao? = null
    @Volatile private var loadJob: Job? = null

    /**
     * In-memory mirror so Compose can resolve a name synchronously in a composition — Room
     * would otherwise be a disk hit on every row, and blocking queries are illegal on the
     * main thread anyway. Rebuilt from the database on launch and kept in step by every
     * upsert below; the database stays authoritative.
     *
     * A [SnapshotStateMap] so a contacts sync recomposes the screens showing those names.
     */
    private val byId: SnapshotStateMap<String, UserRow> = mutableStateMapOf()

    /** Idempotent. Safe (and cheap) to call from any entry point — Activity, FCM, service. */
    fun init(context: Context) {
        if (dao == null) {
            synchronized(this) {
                if (dao == null) {
                    dao = VoiidDatabase.get(context).users()
                    loadJob = scope.launch { reload() }
                }
            }
        }
    }

    /**
     * Init AND wait for the first load. For the push path only: a cold process woken by a
     * ring push has microseconds-old state and must not render the caller id while the
     * mirror is still empty.
     */
    suspend fun ready(context: Context) {
        init(context)
        loadJob?.join()
    }

    /**
     * Sign-out teardown only (see [com.voiid.app.net.SessionTeardown]): drops the in-memory
     * mirror of every saved contact name, phone number and photo. The `users` TABLE itself is
     * truncated separately by [com.voiid.app.store.VoiidDatabase.wipeAllForSignOut] (one
     * transaction for every table) — this only has to clear the mirror so nothing here
     * re-populates it via [reload] in the gap before that runs.
     */
    fun wipe() {
        byId.clear()
    }

    // MARK: - Reads

    /** Null only for a user we have never seen. Callers must NOT fall back to the raw id. */
    fun user(userId: String): UserRow? = byId[userId]

    /**
     * The name for a user id, with the full precedence chain applied.
     *
     * [fallback] covers the genuinely-unknown case (a call from someone not in the directory
     * yet). Pass a conversation title if you have one; pass nothing and the user sees
     * "Unknown", which is still better than a UUID. A fallback that IS the user id is
     * ignored — that is exactly the bug being fixed.
     */
    fun displayName(userId: String, fallback: String? = null): String {
        byId[userId]?.let { return it.displayName() }
        val f = fallback?.trim()
        if (!f.isNullOrEmpty() && f != userId) return f
        return "Unknown"
    }

    fun photoUrl(userId: String): String? = byId[userId]?.photoUrl

    /**
     * The peer's number in strict E.164, or null.
     *
     * Null is the NORMAL case for a peer met by username, or one we have not seen in a
     * `contacts/discover` pass yet: `phone_e164` is only ever written by the address-book
     * path ([upsertFromAddressBook] / [upsertManyFromAddressBook]) because the backend
     * matches contacts by SHA-256 hash and never returns raw numbers. Callers must treat
     * null as "no tel: handle for this person" and fall back, exactly as iOS falls back
     * to a `.generic` CXHandle.
     *
     * Normalized on the way out as well as on the way in — see [com.voiid.app.net.E164].
     */
    fun phoneE164(userId: String): String? =
        com.voiid.app.net.E164.normalize(byId[userId]?.phoneE164)

    /**
     * Reverse lookup: a phone number the SYSTEM handed us -> our user id. Direct port of
     * `CallIntentRouter.userId(forPhone:)`.
     *
     * This is what turns a tap on a contact-card row, or a redial from the system call log,
     * back into a Voiid call: the OS has no idea what a Voiid user id is, so the only thing
     * it can give us back is the number we put on the handle in the first place.
     *
     * Compares NORMALIZED numbers on BOTH sides. The system returns whatever string it
     * stored — which may be spaced, dashed, or re-formatted by the dialer — so a raw string
     * comparison would silently fail to find a user we plainly know.
     *
     * Reads the in-memory mirror only, so it is safe on the main thread. A cold process
     * (push- or intent-woken) should [ready] first; see [userIdForPhoneBlocking] for the
     * path that cannot.
     */
    fun userIdForPhone(raw: String?): String? {
        val target = com.voiid.app.net.E164.normalize(raw) ?: return null
        return byId.entries.firstOrNull {
            com.voiid.app.net.E164.normalize(it.value.phoneE164) == target
        }?.key
    }

    /**
     * [userIdForPhone] for a cold, intent-woken process: hits the database directly rather
     * than trusting a mirror that may not have loaded yet. IO — never call from the main
     * thread. Still normalizes on both sides, because the stored column was written by the
     * discovery normalizer and is not guaranteed to be strict E.164.
     */
    fun userIdForPhoneBlocking(raw: String?): String? {
        userIdForPhone(raw)?.let { return it }
        val target = com.voiid.app.net.E164.normalize(raw) ?: return null
        val d = dao ?: return null
        val rows = runCatching { d.withPhone() }.getOrDefault(emptyList())
        return rows.firstOrNull { com.voiid.app.net.E164.normalize(it.phoneE164) == target }?.userId
    }

    /**
     * Every peer we hold a usable number for. The contacts sync adapter's input set: a peer
     * with no number cannot be merged into an address-book entry and must not be written at
     * all, or it lands as an orphan "Voiid" contact.
     */
    fun withPhoneNumbers(): List<UserRow> =
        byId.values.filter { com.voiid.app.net.E164.normalize(it.phoneE164) != null }

    /** [withPhoneNumbers] straight from the database — for the sync adapter's own thread. */
    fun withPhoneNumbersBlocking(): List<UserRow> =
        withPhoneNumbersBlockingOrNull() ?: emptyList()

    /**
     * Like [withPhoneNumbersBlocking] but returns `null` when the DATABASE READ FAILED, as
     * distinct from an empty list meaning "this user genuinely has no Voiid contacts".
     *
     * The distinction is load-bearing for the contacts sync: it deletes every raw contact
     * NOT in this list, so treating a transient Room failure as "no peers" would wipe every
     * Voiid call row from the user's address book (to be recreated on the next good sync — a
     * visible flap). The sync adapter must abort its destructive pass on null.
     */
    fun withPhoneNumbersBlockingOrNull(): List<UserRow>? {
        val d = dao ?: return withPhoneNumbers()
        val rows = runCatching { d.withPhone() }.getOrNull() ?: return null
        return rows.filter { com.voiid.app.net.E164.normalize(it.phoneE164) != null }
    }

    /** Re-read the whole table into the mirror. Cheap: this table is contact-sized. */
    private suspend fun reload() {
        val d = dao ?: return
        val rows = withContext(Dispatchers.IO) { runCatching { d.all() }.getOrDefault(emptyList()) }
        val fresh = rows.associateBy { it.userId }
        // Diff-update instead of clear-then-refill. This method runs on a background
        // dispatcher after EVERY upsert while Compose readers and the call UI resolve names
        // off `byId` concurrently; a `clear()` opened a window where every lookup returned
        // "Unknown". Removing only the genuinely-departed keys and putting the rest never
        // empties the map, so a name in the table is resolvable throughout the refresh.
        val stale = byId.keys - fresh.keys
        for (k in stale) byId.remove(k)
        byId.putAll(fresh)
    }

    // MARK: - Writes
    //
    // Each source updates ONLY the columns it is authoritative for (see UserDao). That is the
    // whole point of storing the name components separately: a profile refresh from the server
    // must never clobber the address-book name, and a contacts re-sync must never clobber the
    // peer's own profile fields.

    /** From the device address book. Authoritative for `saved_name` + the number, nothing else. */
    fun upsertFromAddressBook(userId: String, savedName: String?, phoneE164: String?) {
        val d = dao ?: return
        scope.launch {
            d.upsertFromAddressBook(userId, savedName.orNull(), phoneE164.orNull(), nowSeconds())
            reload()
        }
    }

    /** From a server profile payload. Authoritative for the peer's self-set fields. */
    fun upsertFromServer(
        userId: String,
        fullName: String? = null,
        username: String? = null,
        photoUrl: String? = null,
        bio: String? = null,
        phoneE164: String? = null,
    ) {
        val d = dao ?: return
        scope.launch {
            d.upsertFromServer(
                userId, fullName.orNull(), username.orNull(), photoUrl.orNull(),
                bio.orNull(), phoneE164.orNull(), nowSeconds(),
            )
            reload()
        }
    }

    /**
     * Bulk variant for a contacts sync — one transaction and one mirror rebuild, so a
     * 500-contact refresh doesn't recompose every observing screen 500 times.
     */
    fun upsertManyFromAddressBook(entries: List<Triple<String, String?, String?>>) {
        if (entries.isEmpty()) return
        val d = dao ?: return
        scope.launch {
            d.upsertManyFromAddressBook(
                entries.map { Triple(it.first, it.second.orNull(), it.third.orNull()) },
                nowSeconds(),
            )
            reload()
        }
    }

    private fun nowSeconds() = System.currentTimeMillis() / 1000

    /** Blank is "I know nothing about this column", not "clear it" — see the COALESCE merges. */
    private fun String?.orNull(): String? = this?.trim()?.takeIf { it.isNotEmpty() }
}

/** The name to show for this row. See the precedence list on [UserDirectory]. */
fun UserRow.displayName(): String {
    for (candidate in listOf(savedName, fullName, phoneE164, username)) {
        val c = candidate?.trim()
        if (!c.isNullOrEmpty()) return c
    }
    return "Unknown"
}

File: apps/ios/Voiid/Voiid/Storage/UserDirectory.swift
//
//  UserDirectory.swift
//  Voiid
//
//  The ONE place a human-readable name for a user id is produced.
//
//  WHY THIS EXISTS: names were resolved ad hoc at ~15 call sites, each with its own
//  fallback, and the call paths had no fallback at all — so an incoming call showed a
//  raw user UUID, and chat screens showed whatever name the peer typed at signup even
//  when you had them saved as something else in your address book. The address book
//  WAS read correctly (ContactsService), but only into a 120s in-memory cache that
//  died on relaunch and was never consulted by the call or chat UI.
//
//  PRECEDENCE (applies everywhere — calls, chat list, chat header, profile):
//     1. saved_name   — this device's address book. If you saved them as "Mum",
//                       every screen says "Mum", whatever they call themselves.
//     2. full_name    — the peer's own profile name.
//     3. phone_e164   — a number is still useful.
//     4. username     — Clips handle, last resort.
//     5. "Unknown"    — NEVER a raw user id. A UUID on screen is always a bug.
//
//  PRIVACY NOTE: this is also why the ring push carries no caller name. Sending one
//  would tell Apple and Google who is calling whom; instead the callee resolves the
//  name locally from `caller_id` through this directory, which needs no network and
//  works when the app was just launched cold by a push.
//

import Combine
import Foundation
import GRDB

/// A person as the app knows them locally, merged from every source.
struct DirectoryUser: Codable, FetchableRecord, PersistableRecord, Identifiable {
    static let databaseTableName = "users"

    var userId: String
    var savedName: String?
    var fullName: String?
    var username: String?
    var phoneE164: String?
    var photoURL: String?
    var bio: String?
    var updatedAt: Int64

    var id: String { userId }

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case savedName = "saved_name"
        case fullName = "full_name"
        case username
        case phoneE164 = "phone_e164"
        case photoURL = "photo_url"
        case bio
        case updatedAt = "updated_at"
    }

    /// The name to show. See the precedence list at the top of this file.
    var displayName: String {
        for candidate in [savedName, fullName, phoneE164, username] {
            if let c = candidate?.trimmingCharacters(in: .whitespacesAndNewlines), !c.isEmpty {
                return c
            }
        }
        return "Unknown"
    }
}

@MainActor
final class UserDirectory: ObservableObject {
    static let shared = UserDirectory()

    private let db = VoiidDatabase.shared

    /// In-memory mirror so SwiftUI bodies can resolve a name synchronously without
    /// touching SQLite on every row render. Rebuilt from the database on launch, and
    /// kept in step by every upsert below — the database stays authoritative.
    @Published private(set) var byId: [String: DirectoryUser] = [:]

    private init() { reload() }

    // MARK: - Reads

    /// Synchronous, non-throwing, safe in a SwiftUI `body`. Returns nil only for a
    /// user we have never seen — callers decide what to show, but must never fall
    /// back to the raw id.
    func user(_ userId: String) -> DirectoryUser? { byId[userId] }

    /// The name for a user id, with the full precedence chain applied.
    ///
    /// `fallback` covers the genuinely-unknown case (a call from someone not in the
    /// directory yet). Pass a phone number or a group title if you have one; pass
    /// nothing and the user sees "Unknown", which is still better than a UUID.
    func displayName(_ userId: String, fallback: String? = nil) -> String {
        if let u = byId[userId] { return u.displayName }
        if let f = fallback?.trimmingCharacters(in: .whitespacesAndNewlines),
           !f.isEmpty, f != userId {
            return f
        }
        return "Unknown"
    }

    func photoURL(_ userId: String) -> String? { byId[userId]?.photoURL }

    /// Everyone a story can actually REACH: the address-book directory UNION every 1:1
    /// conversation peer.
    ///
    /// The directory alone is the wrong set. It is populated by address-book matching
    /// (upsertFromAddressBook) and a few profile fetches, so someone you chat with every day
    /// but never SAVED as a contact is absent from it — and the story composer, which built
    /// its audience from `byId` alone, could not even offer them. Their story never reached
    /// that person, and the reason was invisible.
    ///
    /// Android has always used conversation peers (StoriesStore.candidateAudience), so this
    /// also removes a real cross-platform asymmetry in who can receive a story. The union is
    /// the honest answer: a story ride needs an established 1:1 session, which either source
    /// implies. Ids with no directory row still resolve a name via `displayName(_:)`.
    func storyReachableUserIds() -> Set<String> {
        let me = TokenStore.shared.userId
        var ids = Set(byId.keys)
        for c in LocalStore.conversations() where c.type == .direct {
            if let peer = c.peerUserId, !peer.isEmpty { ids.insert(peer) }
        }
        if let me { ids.remove(me) }
        return ids
    }

    /// Forget everyone. Called on sign-out: the directory holds address-book names, phone
    /// numbers and photos for the previous account's contacts, and the process does not
    /// restart, so without this the next user's call screens and chat headers would resolve
    /// names from the previous user's address book.
    func wipe() {
        byId.removeAll()
    }

    func reload() {
        let rows = db.read { db in try DirectoryUser.fetchAll(db) } ?? []
        byId = Dictionary(uniqueKeysWithValues: rows.map { ($0.userId, $0) })
    }

    // MARK: - Writes
    //
    // Each source updates ONLY the columns it is authoritative for. This is the whole
    // point of storing the name components separately: a profile refresh from the
    // server must never clobber the address-book name, and a contacts re-sync must
    // never clobber the peer's own profile fields.

    /// From the device address book (ContactsService discovery). Authoritative for
    /// `saved_name` and the phone number, and for nothing else.
    ///
    /// The number is normalized to strict E.164 on the way in. CallKit matches a call
    /// to an address-book contact by the handle STRING, so storing "+91 91234 56789"
    /// here and "+919123456789" there would give one person two separate entries in
    /// the phone app's Recents. Canonical on write, canonical everywhere.
    func upsertFromAddressBook(userId: String, savedName: String?, phoneE164: String?) {
        let phoneE164 = CallManager.normalizedE164(phoneE164) ?? phoneE164
        db.write { db in
            try db.execute(sql: """
                INSERT INTO users (user_id, saved_name, phone_e164, updated_at)
                VALUES (?, ?, ?, ?)
                ON CONFLICT(user_id) DO UPDATE SET
                    saved_name = COALESCE(excluded.saved_name, users.saved_name),
                    phone_e164 = COALESCE(excluded.phone_e164, users.phone_e164),
                    updated_at = excluded.updated_at
                """, arguments: [userId, savedName, phoneE164, Int64(Date().timeIntervalSince1970)])
        }
        reload()
    }

    /// From a server profile payload. Authoritative for the peer's self-set fields.
    func upsertFromServer(userId: String, fullName: String?, username: String?,
                          photoURL: String?, bio: String? = nil, phoneE164: String? = nil) {
        db.write { db in
            try db.execute(sql: """
                INSERT INTO users (user_id, full_name, username, photo_url, bio, phone_e164, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(user_id) DO UPDATE SET
                    full_name  = COALESCE(excluded.full_name,  users.full_name),
                    username   = COALESCE(excluded.username,   users.username),
                    photo_url  = COALESCE(excluded.photo_url,  users.photo_url),
                    bio        = COALESCE(excluded.bio,        users.bio),
                    phone_e164 = COALESCE(excluded.phone_e164, users.phone_e164),
                    updated_at = excluded.updated_at
                """, arguments: [userId, fullName, username, photoURL, bio, phoneE164,
                                 Int64(Date().timeIntervalSince1970)])
        }
        reload()
    }

    /// Bulk variant of `upsertFromServer`. Use this for anything that resolves more
    /// than one peer — a conversation sync calls the single-row version once per
    /// direct chat, and each of those reloads the WHOLE table and publishes to every
    /// observing view. One transaction, one reload.
    func upsertManyFromServer(_ entries: [(userId: String, fullName: String?,
                                          username: String?, photoURL: String?)]) {
        guard !entries.isEmpty else { return }
        let now = Int64(Date().timeIntervalSince1970)
        db.write { db in
            for e in entries {
                try db.execute(sql: """
                    INSERT INTO users (user_id, full_name, username, photo_url, updated_at)
                    VALUES (?, ?, ?, ?, ?)
                    ON CONFLICT(user_id) DO UPDATE SET
                        full_name  = COALESCE(excluded.full_name,  users.full_name),
                        username   = COALESCE(excluded.username,   users.username),
                        photo_url  = COALESCE(excluded.photo_url,  users.photo_url),
                        updated_at = excluded.updated_at
                    """, arguments: [e.userId, e.fullName, e.username, e.photoURL, now])
            }
        }
        reload()
    }

    /// Bulk variant for a contacts sync — one transaction, one reload, so a 500-contact
    /// refresh doesn't publish 500 times and thrash every observing view.
    func upsertManyFromAddressBook(_ entries: [(userId: String, savedName: String?, phoneE164: String?)]) {
        guard !entries.isEmpty else { return }
        let now = Int64(Date().timeIntervalSince1970)
        // Canonical E.164 on write — see upsertFromAddressBook for why this matters
        // to the phone app's call log.
        let entries = entries.map {
            (userId: $0.userId,
             savedName: $0.savedName,
             phoneE164: CallManager.normalizedE164($0.phoneE164) ?? $0.phoneE164)
        }
        db.write { db in
            for e in entries {
                try db.execute(sql: """
                    INSERT INTO users (user_id, saved_name, phone_e164, updated_at)
                    VALUES (?, ?, ?, ?)
                    ON CONFLICT(user_id) DO UPDATE SET
                        saved_name = COALESCE(excluded.saved_name, users.saved_name),
                        phone_e164 = COALESCE(excluded.phone_e164, users.phone_e164),
                        updated_at = excluded.updated_at
                    """, arguments: [e.userId, e.savedName, e.phoneE164, now])
            }
        }
        reload()
    }
}
